### PR TITLE
Set workspace address when attempting to Book Travel

### PR DESCRIPTION
**Pull request by rushatgabhane**
### Fixed Issues

$ https://github.com/Expensify/App/issues/46313
$ https://github.com/Expensify/App/issues/43780

PROPOSAL:

### Test 1 : Address check

1. On account B, create a new workspace and invite account A. 
2. A should not be an admin of the workspace
3. Go to account A
4. Make B's workspace as your default workspace!
5. Green Fab > Book Travel
6. Verify that company address page opens
7. Fill an address and save
8. Click book travel 
9. Verify that you are taken to travelDot 

### Test 2 : Travel terms and condition
1. Go to a new account that has not accepted travel terms of condition
2. Green Fab > Book Travel
3. QA can skip : Make `Travel.acceptSpotnanaTerms()` reject the promise.
4. Verify that when there is an error, travelDot is not opened and you see an error message on terms and condtion.

### Test 3 : Open travel
1.  Green Fab > Book Travel
2. QA can skip : Make `Link.openTravelDotLink()` reject the promise.
3. Verify that when there is an error, travelDot is not opened and you see an error message.

### Test 4 : Verify that production is opened unless NVP is present 
1. Green Fab > Book Travel
2. Book travel
3. Verify that production travelDot is opened
4. Hardcode `NVP.TRAVEL_SETTINGS.testAccount` to  true
5. Verify that staging travelDot is opened

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/a433f125-cb32-4de0-94fe-211c31a97462

https://github.com/user-attachments/assets/5b2c81d3-de47-42de-a789-8364f1b55d9b

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/2889b1dc-7e74-48a6-bfe2-bdc27fb421ac

https://github.com/user-attachments/assets/2b48c67e-01f4-409a-844e-133b3050a022

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/a8f966ec-6af3-4960-af3b-c2e343d36439

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/437edd8d-6ab8-4380-a59b-18bbd4959a6b

### Accept spotnana

https://github.com/user-attachments/assets/80839995-0ef0-4a7e-bc8d-06696c3d5bd9

### Open Travel

https://github.com/user-attachments/assets/c50b8207-1fc8-43cd-bd69-7347c2dc779c

</details>

<details>
<summary>MacOS: Desktop</summary>

image

</details>

### CODE REVIEW

**Comment by rushatgabhane** [comment]
Error on spotnana side is a known issue -

image


https://expensify.slack.com/archives/C05S5EV2JTX/p1722034497413689?thread_ts=1721406118.373969&cid=C05S5EV2JTX

**Comment by rushatgabhane** [comment]
cc @stitesExpensify

File: src/components/ReportActionItem/MoneyRequestView.tsx [hunk 1]
```diff
@@ -621,7 +621,9 @@ function MoneyRequestView({
                         icon={Expensicons.Suitcase}
                         iconRight={Expensicons.NewWindow}
                         shouldShowRightIcon
-                        onPress={() => Link.openTravelDotLink(activePolicyID, CONST.TRIP_ID_PATH(tripID))}
+                        onPress={() => {
+                            Link.openTravelDotLink(activePolicyID, CONST.TRIP_ID_PATH(tripID));
```
[the remark this answers is not in the document]
Comment (rushatgabhane) [reply src/components/ReportActionItem/MoneyRequestView.tsx:625]: > comment for safety

im sorry i dont understand. what comment should i add?

**Comment by rushatgabhane** [comment]
> Should we show option here to set the workspace address? I am not super sure about it

yes the linked issue is to allow the user to set a workspace address even if they're non-admin

File: src/libs/actions/Link.ts [hunk 2]
```diff
@@ -95,13 +104,21 @@ function openTravelDotLink(policyID: OnyxEntry<string>, postLoginPath?: string)
...
-        (travelDotURL) => travelDotURL,
-    );
+    return new Promise((_, reject) => {
+        asyncOpenURL(
+            // eslint-disable-next-line rulesdir/no-api-side-effects-method
+            API.makeRequestWithSideEffects(SIDE_EFFECT_REQUEST_COMMANDS.GENERATE_SPOTNANA_TOKEN, parameters, {})
+                .then((response) => {
+                    if (!response?.spotnanaToken) {
+                        reject(new Error('Failed to generate spotnana token.'));
+                        return;
```
[the remark this answers is not in the document]
Comment (rushatgabhane) [reply src/libs/actions/Link.ts:114]: @shubham1206agra could you please explain why we should throw an error?

File: src/libs/asyncOpenURL/index.ts [hunk 3]
```diff
@@ -7,7 +7,14 @@ const asyncOpenURL: AsyncOpenURL = (promise, url) => {
...
-        Linking.openURL(typeof url === 'string' ? url : url(params));
+        if (typeof url === 'string') {
+            Linking.openURL(url);
+            return;
+        }
+        if (!params) {
+            return;
+        }
+        Linking.openURL(url(params));
     });
```
[the remark this answers is not in the document]
Comment (rushatgabhane) [reply src/libs/asyncOpenURL/index.ts:18]: why? and what should we do when we catch an error?

**Comment by rushatgabhane** [comment]
@shubham1206agra all fixed. Thank you for the review and thorough testing

**Comment by rushatgabhane** [comment]
Ready for re-review @shubham1206agra @stitesExpensify

**Review by shubham1206agra — APPROVED** [review]
I am approving this right now. Only one minor thing left. But this is urgent to merge.

**Comment by rushatgabhane** [comment]
> Can you address this?

it's in issue fesc

**Review by stitesExpensify — DISMISSED** [review]
Looks great! Just one question and then I will merge

File: src/libs/asyncOpenURL/index.ts [hunk 4]
```diff
@@ -6,9 +6,11 @@ const asyncOpenURL: AsyncOpenURL = (promise, url) => {
...
     }
 
-    promise.then((params) => {
-        Linking.openURL(typeof url === 'string' ? url : url(params));
-    });
+    promise
+        .then((params) => {
+            Linking.openURL(typeof url === 'string' ? url : url(params));
+        })
+        .catch(() => {});
```
[the remark this answers is not in the document]
Comment (rushatgabhane) [reply src/libs/asyncOpenURL/index.ts:13]: @stitesExpensify we can log an error for server.

Would a generic message be helpful?
eg: `error occurred while trying to open link: {link}`

[the remark this answers is not in the document]
Comment (stitesExpensify) [reply src/libs/asyncOpenURL/index.ts:13]: Yeah like that!

[the remark this answers is not in the document]
Comment (stitesExpensify) [reply src/libs/asyncOpenURL/index.ts:13]: If you update that now I'll get it merged today :D

**Comment by rushatgabhane** [comment]
@stitesExpensify done

**Review by stitesExpensify — APPROVED** [review]

**Comment by IuliiaHerets** [comment]
This PR is failing because of issue #46971

https://github.com/user-attachments/assets/2a2b8d81-090e-443d-8b28-a47505fd0d32

File: src/pages/Travel/ManageTrips.tsx [hunk 5]
```diff
@@ -49,12 +59,22 @@ function ManageTrips() {
                     ctaText={translate('travel.bookTravel')}
                     ctaAccessibilityLabel={translate('travel.bookTravel')}
                     onCtaPress={() => {
+                        if (!hasPolicyAddress) {
+                            Navigation.navigate(ROUTES.WORKSPACE_PROFILE_ADDRESS.getRoute(activePolicyID ?? '-1'));
```
Comment (rojiphil) [inline src/pages/Travel/ManageTrips.tsx:63]: Not using `backTo` while calling `getRoute` caused #47069

**Pull request merged**